services: Change filter dropdown to a native select

garrett opened the pull request:
Backported the styled native select from #11155 and adjusted the HTML+CSS+JS accordingly.

Originally brought up in #11241.

garrett commented:
The changes probably need tests to be fixed. (Sorry.)

marusak commented:
@garrett does it intentionally look like this? It kinda is too big, too much space around divider, the divider does not span across...
![bigbig](https://user-images.githubusercontent.com/12330670/53963301-52144e00-40ed-11e9-9d66-19961d90337f.png)

garrett commented:
That's the best anyone can do with "dividers" in native dropdowns. It's just a bunch of UTF-8 line drawing characters in a row, in a disabled entry... not an actual divider. We can't have it go all the way across, as we don't know how long things will be, and if it goes too far, it makes the element too wide (as it's roughly sized on the content).

So, yes, that's intended. It's one of the few drawbacks. But there are so many advantages otherwise.

marusak commented on pkg/systemd/services.html:
```diff
@@ -62,19 +62,15 @@
-          </ul>
-        </div>
+
+        <label id="services-dropdown-label" translatable="yes" class="hidden">Status</label>
```
`Status` is not a great label. It should be most likely `Automatic start` or `Automatic start status` or something similar, but lets have there `automatic start` as it really reefers to it (also see the name of the column)

marusak replied:
s/start/startup

garrett replied:
Oh, hey! Here's an idea: We could even use the same string as the label in the `th` of the table for the `aria-labelledby` and drop the hidden `<label>`, as the select is conceptually tied to that column in the list!

Thankfully, `aria-labelledby` still works if the element is hidden (as long as it still exists n the DOM):
https://developer.paciellogroup.com/blog/2015/05/short-note-on-aria-labelledby-and-aria-describedby/

marusak replied:
Sounds good :)

garrett replied:
Oh, we didn't actually implement this change yet. Oops! :grin:

marusak commented on pkg/systemd/services.html:
```diff
@@ -62,19 +62,15 @@
+        <label id="services-dropdown-label" translatable="yes" class="hidden">Status</label>
+        <select class="ct-select" id="services-dropdown" aria-labelledby="services-dropdown-label">
+          <option value="0" translatable="yes" selected>All</option>
+          <option role="separator" className="divider" disabled="disabled">──────────</option>
```
Lets probably drop this divider. It does not look that good, list it rather short, and also screen reader see it as element that is not accessible.

garrett replied:
Yeah. I wonder what other screen readers do too. Since it's the simplest option (ha! 😏) to remove it, this makes perfect sense.

marusak commented:
Also sometimes search and dropdown don't align very nicely.
![notaligned](https://user-images.githubusercontent.com/12330670/53967931-6e68b880-40f6-11e9-8199-a1ebf1ddc0e9.png)

garrett commented:
> Also it seem that space between timer button disapperd with this PR, please check it out.

Oops.

Oh they're aligning by baseline. I guess we can change that, assuming they're the same height (as they're both widgets set to the same height, it should be safe to do so here).

garrett commented:
It's nearing the end of the day (and workweek for me) here, and I'm not done yet. :slightly_frowning_face: I'll have to finish working with you on this next week. :grin:

garrett commented:
I've reworked the header as a grid to be able to preserve source order and still decompose in a decent manner and also work in small sizes. It likely needs a little more additional work.

marusak commented:
> I've reworked the header as a grid to be able to preserve source order and still decompose in a decent manner and also work in small sizes. It likely needs a little more additional work.

Can you please explain the problem you are trying to solve? Before this last commit it looked like this:
![beforepatch](https://user-images.githubusercontent.com/12330670/54340930-d0ff0e80-4638-11e9-8a4c-f63fcfd79cad.png)
and with the last commit it looks like this:
![newbroken](https://user-images.githubusercontent.com/12330670/54340943-d8beb300-4638-11e9-8f09-400a361bdb61.png)
I purposefully broke the top panel that way so the filtering stays in the panel and not jumps out of it. Also it needs to be above the tab selection as it looks very weird if there is something under tabs in the panel.

garrett commented:
Oh, I thought I mentioned the reasons here... but I don't see it. Perhaps it was in a commit, or perhaps I just forgot to click the big green button.

1. Don't break the look of the tabs
2. Keep visual order the same as document order
3. Work in small widths
4. Account for the action button when timers is selected

garrett commented:
To address the list in my previous content, there are various solutions:

1. Keep it as-is as it currently is in this PR
2. Flip the filter and tabs, so tabs are on the right and filter box is on the left
3. Detatch the tabs in mobile mode (so it's more like a list of buttons again)
4. Always have the filter box in the visible content area

@marusak, @andreasn: Which options do you two like the best?

marusak commented:
> 1. Keep it as-is as it currently is in this PR

I don't like it as it is now in this PR - since filters jump out of the panel, it looks rather broken.

> 2. Flip the filter and tabs, so tabs are on the right and filter box is on the left

This is interesting idea, but I would need to see it first. Can you post some screenshot how this looks like on a big and on a small screen?

> 3. Detatch the tabs in mobile mode (so it's more like a list of buttons again)

I am not sure what you mean here - you mean styling it on small screens as group of buttons and let it wrap?

> 4. Always have the filter box in the visible content area

This would be great, but I don't think it solves problem of breaking on small screens.

garrett commented:
>> 2. […] tabs are on the right and filter box is on the left […]
>
> This is interesting idea, but I would need to see it first

Sure thing.

Here's a quick mockup I Frankensteined from your screenshots, using the GIMP:

![flipped](https://user-images.githubusercontent.com/10246/55396819-15e5d900-5545-11e9-8cf1-d334d591aca6.png)

In mobile, the filter bar would be at the top, with the tabs below, but the source order would be the same. And both would be in the floating toolbar still.


> on small screens as group of buttons and let it wrap

Yep! Totally. It basically would revert to a group of buttons, (more or less) like before — but only on mobile.

>> 4. Always have the filter box in the visible content area
>
> This would be great, but I don't think it solves problem of breaking on small screens.

It'd look like it does on mobile, where the tabs are at the top and the filter bar is below. Basically, like this screenshot of yours:

![newbroken](https://user-images.githubusercontent.com/12330670/54340943-d8beb300-4638-11e9-8f09-400a361bdb61.png)

marusak commented:
> Here's a quick mockup I Frankensteined from your screenshots, using the GIMP:

I like it. It looks a bit odd now, but mostly because I am used to the type selector being on left. @andreasn what do you think?
If @andreasn would be against then I would go with buttons group on smaller screen.

garrett commented:
I thought about this a bit more, and especially with longer-stringed languages, the tabs may overflow anyway. I've reverted my grid decomposition changes locally and will re-do the tab group as a set of floating buttons for mobile widths.

The tabs could be either on the left or right, but either way, they will be in source order in both desktop and mobile widths.

garrett commented:
@marusak: I accidentally went overboard and made the list work (mostly) well in mobile too. I should have a WIP PR for this soon. :wink:

I'll need some help making the list work _truly_ well, however — as it refreshes the whole thing and then you lose your place when scrolling around. But the HTML & CSS are mostly done, and it's at least much better than it currently is.

garrett commented:
@marusak, @andreasn: If you two could check out this PR and test it on a mobile device, that would be awesome. I'd like to hear feedback from both of you one this.

I think this is one of the ways we could solve our lists working well on mobile.

(FWIW: This is similar to what Wikipedia does for tables.)

Ideally, I'd make this a bit more generic and just add it to our lists (or at least do so as a class).

marusak commented:
The buttons look very nice :) I think this approach is the most robust  and still looks pleasing.
> I'll need some help making the list work truly well

If I can help you somehow, just let me know

Screenshots if anyone is interested but not enough to build it :P 
![bigger](https://user-images.githubusercontent.com/12330670/55861913-a93b9180-5b77-11e9-9636-068bb30915af.png)
![smaller](https://user-images.githubusercontent.com/12330670/55861915-a93b9180-5b77-11e9-8730-a0a118a96f51.png)
![smallphone](https://user-images.githubusercontent.com/12330670/55861911-a8a2fb00-5b77-11e9-866d-2af4e83c249b.png)

andreasn commented:
Works for me!

garrett commented:
I've optimized touch in a few ways and also have been struggling with Chrome on Android, hitting some esoteric bug that none of the docs reference (yet again) and is (apparently) related to an open bug on Chrome's bug tracker that's been there for a decade already. :roll_eyes:

Chrome really is the new IE. :nauseated_face: 

I've implemented the overscroll in a method using the background with multiple backgrounds instead of using image masking (which _works_, but the `position: fixed` in Chrome on Android is broken, so that method would need tweaking further somehow with some workaround, if possible...).

I'd like to see if this new method could use an overlay, so we can also affect the text. Currently, it's just background colors that matches the divider lines (so they don't stand out).

---

Also, if we get this successful enough, I'd like to base a more generic way to overflow lists for mobile for Cockpit based on this technique.

Ideally, we'd have 3 methods for mobile lists on Cockpit (and all 3 would be used):
1. Overflow (basically this method)
2. Hiding unnecessary content (in the case of the Software page, where duplicated content is hidden)
3. Deconstruction (which is tricky — but not impossible — to get right and have accessibility)

Anyway, I spent time today trying to get this right due to making this a candidate for method 1.

---

This all said, if you two could review the PR in its current state on your phones, that would be great!

garrett commented:
I've optimized touch in a few ways and also have been struggling with Chrome on Android, hitting some esoteric bug that none of the docs reference (yet again) and is (apparently) related to an open bug on Chrome's bug tracker that's been there for a decade already. :roll_eyes:

Chrome really is the new IE. :nauseated_face: 

I've implemented the overscroll in a method using the background with multiple backgrounds instead of using image masking (which _works_, but the `position: fixed` in Chrome on Android is broken, so that method would need tweaking further somehow with some workaround, if possible...).

I'd like to see if this new method could use an overlay, so we can also affect the text. Currently, it's just background colors that matches the divider lines (so they don't stand out).

---

Also, if we get this successful enough, I'd like to base a more generic way to overflow lists for mobile for Cockpit based on this technique.

Ideally, we'd have 3 methods for mobile lists on Cockpit (and all 3 would be used):
1. Overflow (basically this method)
2. Hiding unnecessary content (in the case of the Software page, where duplicated content is hidden)
3. Deconstruction (which is tricky — but not impossible — to get right and have accessibility)

Anyway, I spent time today trying to get this right due to making this a candidate for method 1.

---

This all said, if you two could review the PR in its current state on your phones, that would be great!

garrett commented:
We might also want to split out the mobile list stuff into another PR, as to not block the original intent of this PR any longer. However, as it touches some of the same code, it would depend on this PR.

andreasn commented:
Looks great on desktop and generally on Mobile, but it seems the dropdown has a disabled radiobutton next to the divider on both Firefox and Chrome. Is it supposed to do that?
![dropdown2](https://user-images.githubusercontent.com/209895/56031972-2d7d4880-5d21-11e9-94ac-5bca2afafe3c.jpg)
![dropdown1](https://user-images.githubusercontent.com/209895/56031973-2d7d4880-5d21-11e9-9d1c-d8be1da0f799.jpg)

andreasn commented:
I have not been able to find a single example of a dropdown with a divider that works well on mobile, so I think we're good too?

andreasn commented:
Dropdown on Firefox and Chrome on Android:
![dropdown1](https://user-images.githubusercontent.com/209895/56033731-733c1000-5d25-11e9-8261-a6c3eb08b1ae.jpg)
![dropdown2](https://user-images.githubusercontent.com/209895/56033732-733c1000-5d25-11e9-8f59-1b1a5b3be401.jpg)

I think the custom divider makes it look a bit weird.
Maybe we should just drop the divider, since if we're going native html, we should go native html full on.

garrett commented:
> I think the custom divider makes it look a bit weird.
> Maybe we should just drop the divider, since if we're going native html, we should go native html full on.

FWIW: Native HTML has no divider. Everybody has to add them in like this.

The alternative is to have groups, which native selectors do have. However, it doesn't make sense here.

But the list is short, all is the default, and the rest are just subsets of all here — they're not really so different — so dropping it here could work.

garrett commented:
I've squashed everything down to two commits and split the mobile-related parts that were in this PR off to #11606. This PR should now only contain the native select changes.

garrett commented:
Oops. I lost the margin on the timer button.

So I squashed in a couple fixups and force-pushed, addressing:
1. Margin on the timer button
2. Changed aria-labbeledby to point to the corresponding `<th>` instead of a hidden label.

garrett commented:
@marusak: Good catch. I'd _prefer_ the tabbing be contained to the `<a>`, as it's a typical native element to tab to. But, yeah, we can fix that in the other PR.

garrett commented:
@andreasn: Divider removed. Can you re-review please? :wink:

garrett commented:
@marusak: I did have to redo the patch for tests, but manually applied the changes that were previously in this PR... so it _should_ be the same. Could you confirm the tests are fine? I see a lot of tests failing (at least 1 in every distro) and want to make sure they're just flakes.

marusak commented:
> Could you confirm the tests are fine?

I retried them and the ones that are still red are unrelated flakes.

marusak commented:
Ah, I already approved, so if @andreasn approves as well we can land this

martinpitt commented:
@andreasn, can you please re-review to unblock this?

andreasn commented:
yup, sorry for the delay!
